runtime/nvqir/CircuitSimulator.h

```diff
@@ -884,6 +890,51 @@ class CircuitSimulatorBase : public CircuitSimulator {
     return qubits;
   }
 
+  /// @brief Allocate `count` qubits in a specific state.
+  std::vector<std::size_t>
+  allocateQubits(std::size_t count,
+                 const cudaq::SimulationState *state) override {
+    if (!state)
+      return allocateQubits(count);
+
+    if (count != state->getNumQubits())
+      throw std::invalid_argument("Dimension mismatch: the input state doesn't "
+                                  "match the number of qubits");
+
+    std::vector<std::size_t> qubits;
+    for (std::size_t i = 0; i < count; i++)
+      qubits.emplace_back(tracker.getNextIndex());
+
+    if (isInBatchMode()) {
+      // Store the current number of qubits requested
+      batchModeCurrentNumQubits += count;
+
+      // We have an allocated state, it has been set to |0>,
+      // we want to reuse it as is. If the state needs to grow, then
+      // we will ask the subtype to add more qubits.
+      if (qubits.back() < nQubitsAllocated)
+        count = 0;
+      else
+        count = qubits.back() + 1 - nQubitsAllocated;
+    }
+
+    cudaq::info("Allocating {} new qubits.", count);
+
+    previousStateDimension = stateDimension;
+    nQubitsAllocated += count;
+    stateDimension = calculateStateDim(nQubitsAllocated);
+
+    // Tell the subtype to allocate more qubits
+    addQubitsToState(*state);
```

Comment:
Is `state` always freed in this call?

Reply:
We don't free the `state` in this call. It is passed as `const cudaq::SimulationState &` and the simulator will make its copy of the state. The lifetime is managed by the `shared_ptr` data member of `cudaq::state`.

Reply:
OK. It looked like it was being freed, but I might have been looking in the wrong place.